x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/NodeLoadDetector.java

```diff
@@ -98,6 +104,20 @@ private void updateLoadGivenTasks(NodeLoad.Builder nodeLoad, PersistentTasksCust
         }
     }
 
+    private void updateLoadGivenModelAllocations(NodeLoad.Builder nodeLoad, TrainedModelAllocationMetadata trainedModelAllocationMetadata) {
+        if (trainedModelAllocationMetadata != null && trainedModelAllocationMetadata.modelAllocations().isEmpty() == false) {
+            for (TrainedModelAllocation allocation : trainedModelAllocationMetadata.modelAllocations().values()) {
+                if (Optional.ofNullable(allocation.getNodeRoutingTable().get(nodeLoad.getNodeId()))
+                    .map(RoutingStateAndReason::getState)
+                    .orElse(RoutingState.STOPPED)
+                    .consumesMemory()) {
+                    nodeLoad.incAssignedJobMemory(allocation.getTaskParams().estimateMemoryUsageBytes());
+                    nodeLoad.incNumAssignedJobs();
```

Comment:
Similar logic to before, if consumes memory, increment these counts.

One question I have, should we count `STARTING` to `numAllocatingJobs`?

Reply:
I think the question for both `numAssignedJobs` and `numAllocatingJobs` is whether trained models are "jobs".

The settings that control how many you can have are `xpack.ml.max_open_jobs` and  `xpack.ml.node_concurrent_job_allocations`, both of which contain the word "job".

We could update the docs to say that a single allocation of a trained model to a node counts as a "job".

But in terms of what this is doing:

1. If a user has said they want a maximum of 20 "jobs" per node, would 5 data frame analytics jobs, 15 anomaly detection jobs and 1 trained model allocation be OK?
2. If a user has said they want no more than 2 "jobs" allocating simultaneously on a node, would it be acceptable to have 10 data frame analytics jobs, 0 anomaly detection jobs and 2 trained models starting up at the same time on the node? *

(* The interesting thing here is that data frame analytics jobs don't count.)

I am thinking trained models _shouldn't_ be "jobs" for any purpose:

1. We've decided that we'll require memory information be available for assigning ML native stuff to nodes, so the max count is much less important than it used to be when we used to sometimes fall back to it for assignment decisions.
2. We've said in a different discussion (see #75411) that we want to have a limit of 1 PyTorch model being loaded per node at any time.  So counting PyTorch models in with anomaly detection jobs for comparison against `xpack.ml.node_concurrent_job_allocations` isn't helpful.
3. It will be confusing in the docs to have to say that an allocated trained model counts as a "job" sometimes.

This makes me think:

1. We should _not_ call `nodeLoad.incNumAssignedJobs()` for trained models.
2. We should not increment `numAllocatingJobs` for `STARTING` trained models.
3. We should instead have separate counters for `numAllocatedTrainedModels` and `numStartingTrainedModels`.

Reply:
I agree with Dave here.